[Executorch] Add quantized kv cache to oss ci

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #6997
* #7038
* #6996

Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

kimishpatel committed:
[Executorch] Add quantized kv cache to oss ci

Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

larryliu0820 commented on examples/models/llama/source_transformation/quantized_kv_cache.py:
```diff
@@ -10,9 +10,36 @@
+    # Ideally package is installed in only one location but usage of
+    # PYATHONPATH can result in multiple locations.
+    # ATM this is mainly used in CI for qnn runner. Will need to revisit this
+    executorch_package_path = executorch.__path__[-1]
+    libs = list(
+        glob.glob(
+            f"{executorch_package_path}/**/libquantized_ops_aot_lib.*", recursive=True
+        )
+    )
+    assert len(libs) == 1, f"Expected 1 library but got {len(libs)}"
+    logging.info(f"Loading custom ops library: {libs[0]}")
+    torch.ops.load_library(libs[0])
+    op = torch.ops.quantized_decomposed.quantize_per_token.out
+    assert op is not None
+
```
Is this needed? I feel like we can add this to `executorch/kernels/quantized/__init__.py`

kimishpatel replied:
Oh I didnt realize that this was already there in quantized/__init__.py. I can try to patch that one and see if it works. Althought it does feel quite weird, TBH, to have executorch/kernels/quantized to be exposed as python package. But it does contain AOT component, so seems fair.

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add quantized kv cache to oss ci"


Fixes to make sure quantized kv cache works in oss

Differential Revision: [D66269487](https://our.internmc.facebook.com/intern/diff/D66269487/)

[ghstack-poisoned]
